### ISSUE TITLE
the detailed location of the change on "support taking a pair of parallel sentences at once"and "track two different losses "

### ISSUE THREAD

**Issue by livehappy1** [dannigt/mid-align#4]
Hello,I want to know the precise location you did modification in the two files or describe the concrete way to change .Thank you!

**Comment by dannigt** [comment]
Hi, thanks for the interest in this work! 
You could see the exact modifications by running a diff against the original HF TF v4.43.4 (at commit `7e5d46`):
```
git clone https://github.com/dannigt/mid-align.git
git clone https://github.com/huggingface/transformers.git transformers_v4.43.4
cd transformers_v4.43.4
git checkout 7e5d46
cd ..
# diff in tokenization_utils_base.py 
diff ./mid-align/src/transformers/tokenization_utils_base.py ./transformers_v4.43.4/src/transformers/tokenization_utils_base.py
# check all changes 
diff ./mid-align/src/transformers ./transformers_v4.43.4
```
Hope this helps!

**Comment by livehappy1** [comment]
thanks for your reply.In the modeling_llama.py,I was confused about the following code:
&nbsp; loss= task_loss+contrastive_loss.
Why plus contrastive loss?It seems useless.








 ----------Reply to Message----------
 On Wed, Jan 21, 2026 18:33 PM ***@***.***&gt; wrote:

  dannigt left a comment (dannigt/mid-align#4)
 
Hi, thanks for the interest in this work!
 You could see the exact modifications by running a diff against the original HF TF v4.43.4 (at commit 7e5d46):
 git clone https://github.com/dannigt/mid-align.git git clone https://github.com/huggingface/transformers.git transformers_v4.43.4 cd transformers_v4.43.4 git checkout 7e5d46 cd .. # diff in tokenization_utils_base.py  diff ./mid-align/src/transformers/tokenization_utils_base.py ./transformers_v4.43.4/src/transformers/tokenization_utils_base.py # check all changes  diff ./mid-align/src/transformers ./transformers_v4.43.4  
Hope this helps!

**Comment by dannigt** [comment]
Thanks for the follow up. Do you mean [this line](https://github.com/dannigt/mid-align/blob/master/src/transformers/models/llama/modeling_llama.py#L1363C20-L1363C49)?
The implementation also supports jointly training with both the task loss and contrastive loss. The total loss is therefore the sum.

**Comment by livehappy1** [comment]
Hello,I want to know how&nbsp; you achieve the goal of tracking two losses.








 ----------Reply to Message----------
 On Thu, Jan 22, 2026 17:54 PM ***@***.***&gt; wrote:

  dannigt left a comment (dannigt/mid-align#4)
 
Thanks for the follow up. Do you mean this line?
 The implementation also supports jointly training with both the task loss and contrastive loss. The total loss is therefore the sum.

**Comment by dannigt** [comment]
Hi, thanks for the follow up. Tracking multiple losses is supported by https://pypi.org/project/hf-mtask-trainer
The trainer inherits from `HfMultiTaskTrainer` ([here](https://github.com/dannigt/mid-align/blob/c3fa8efec63f7bd8b0757280a6f89d96fc4060a8/src/transformers/trainer.py#L4739))

**Comment by livehappy1** [comment]
thank you.Ang in your paper,I want to know your result based on the ＂no gradient merging",just perform two tasks alternatively ，right?


And I also have another small problem.In your code,why did you set two tasks both true before evaluation on dev dataset?


 ----------Reply to Message----------
 On Mon, Jan 26, 2026 22:45 PM ***@***.***&gt; wrote:

  dannigt left a comment (dannigt/mid-align#4)
 
Hi, thanks for the follow up. Tracking multiple losses is supported by https://pypi.org/project/hf-mtask-trainer
 The trainer inherits from HfMultiTaskTrainer (here)

**Comment by dannigt** [comment]
Thank you for your patience. 

Re gradient merging: Only [this branch](https://github.com/dannigt/mid-align/blob/master/src/transformers/trainer.py#L5056) under `if not do_gradient_merging` is activated. Therefore the implementation for projecting conflicting gradients etc. was not used.

When revisiting the code, I noticed the current implementation only correctly alternates when `gradient_accumulation_steps=1` . For a correct implementation, the toggle in [this line](https://github.com/dannigt/mid-align/blob/master/src/transformers/trainer.py#L5060) should be moved to [here](https://github.com/dannigt/mid-align/blob/master/src/transformers/trainer.py#L5101), to occur every time we do gradient accumulation (rather than every forward pass).

After fixing this, the results did not differ much from reported in the paper (row 3 and 6 of Table 2):

#### Slot Filling Results (F1)

| Model | Supervised | Transfer (15) | Transfer (aligned) |
|-------|---------|---------|---------|
| Paper - llama | 77.0 | 61.7 | 55.5 |
| After fix | 77.1 | 63.4 | 58.0 |
| Paper - qwen | 77.0 | 55.3 | 46.5 |
| After fix | 77.1 | 55.9 | 49.5 |

#### Machine Translation Results (COMET)

| Model | Supervised  | x-->En | En-->x |
|-------|---------|---------|---------|
| Paper - llama | 81.5 | 83.0 | 80.7 |
| After fix | 81.5 | 83.1 | 80.3 |
| Paper - qwen | 77.6 | 82.7 | 76.9 |
| After fix | 78.9 | 82.8 | 77.1 |

I will make a remark on this in the README and add a new branch with this fix.


----

Re eval loop: with the current implementaiton, without setting both tasks to true, one task would be evaluated on half of the dev set, while another task on the other half.  Please see a related discussion in [another issue](https://github.com/dannigt/mid-align/issues/3).
